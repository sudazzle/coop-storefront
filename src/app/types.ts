export enum SortingOrder {
  ASC = 'asc',
  DESC = 'desc'
}

export type OpeningHour = {
  Day: number
  Open: string
  Close: string
}

export type Store = {
  StoreId: string
  Name: string
  Chain: string
  ChainClassName: string
  ChainId: string
  IsEcommerce: boolean
  NewspaperUrl: string
  ChainImage: string
  InStoreServices: string[]
  Lat: number
  Lng: number
  OpeningHoursToday: string
  OpenNow: boolean
  Phone: string
  Address: string
  City: string
  Distance: number
  Email: string
  SLag: string
  OrganizationNumber: string
  SpecialOpeningHours: string[]
  OpeningHours: OpeningHour[]
  AdditionalInformation: unknown
}

export type StoreByLocation = {
  Location: { Bounds: null, Longitude: number, Latitude: number }
  Stores: Store[]
  AdditionalStores: Store[]
  InfoMessage?: string
}