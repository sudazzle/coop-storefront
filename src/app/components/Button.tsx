import React from 'react';

type ButtonProps = React.ButtonHTMLAttributes<HTMLButtonElement> & {
  active?: boolean
}

export const Button: React.FC<ButtonProps> = ({ active, ...rest }) => {
  let className = "hover:bg-blue-900 ease-in duration-300 hover:scale-105 hover:border-blue-900 hover:text-white border border-blue-300 rounded-xl py-2 px-4"
  
  if (active === true) {
    className += " bg-blue-900 text-white"
  } else {
    className += " bg-white text-blue-900"
  }

  return (
    <button className={className} {...rest} />
  );
}