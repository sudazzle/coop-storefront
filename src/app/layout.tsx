import React from 'react';
import type { Metadata } from 'next';
import './globals.css';
import { Logo } from './icons/logo';
import { CoopMarked } from './icons/CoopMarked';
import { CoopMega } from './icons/CoopMega';
import { Prix } from './icons/Prix';
import { Matkroken } from './icons/Matkroken';
import { Obs } from './icons/Obs';
import { Extra } from './icons/Extra';
import { Byggprix } from './icons/Byggprix';
import { ObsBygg } from './icons/ObsBygg';
import { CoopElectro } from './icons/CoopElectro';

export const metadata: Metadata = {
  title: "Butikker - Coop",
};

const StoreLogoWrapper: React.FC<React.HTMLAttributes<HTMLDivElement>> = ({ children, ...props }) => {
  return (
    <div className="w-10 store-logo" {...props}>
      {children}
    </div>
  );
}

export default function RootLayout({
  children,
}: Readonly<{
  children: React.ReactNode;
}>) {
  return (
    <html lang="no">
      <body>
        <div className="flex items-center bg-[#003366] p-5 justify-between">
          <div className="w-24">
            <Logo />
          </div>
          <div className="flex items-center gap-6">
            <ul className="hidden header-breakpoint:flex items-center text-white gap-3">
              <li>Coop.no</li>
              <li>Min Side</li>
              <li>VÃ¥re Butikker</li>
              <li>Medlem</li>
              <li>Coop Kjeder</li>
            </ul>
            <input className="rounded-3xl p-2" type="text" placeholder="Hva leter du etter?" />
          </div>
        </div>
        <div className="flex justify-center gap-0.5 md:gap-3 p-2 md:p-4 bg-[#DCF0FA]">
          <StoreLogoWrapper>
            <CoopMarked />
          </StoreLogoWrapper>
          <StoreLogoWrapper>
            <CoopMega />
          </StoreLogoWrapper>
          <StoreLogoWrapper>
            <Prix />
          </StoreLogoWrapper>
          <StoreLogoWrapper>
            <Matkroken />
          </StoreLogoWrapper>
          <StoreLogoWrapper>
            <Obs />
          </StoreLogoWrapper>
          <StoreLogoWrapper className="w-20">
            <Extra />
          </StoreLogoWrapper>
          <StoreLogoWrapper>
            <Byggprix />
          </StoreLogoWrapper>
          <StoreLogoWrapper className="w-8">
            <ObsBygg />
          </StoreLogoWrapper>
          <StoreLogoWrapper>
            <CoopElectro />
          </StoreLogoWrapper>
        </div>
        { children }
      </body>
    </html>
  );
}
