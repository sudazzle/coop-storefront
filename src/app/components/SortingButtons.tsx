'use client';

import React from 'react';
import { Button } from './Button';
import { SortingOrder } from '@/app/types';

type SortingButtonsProps = {
  onSort: (order: SortingOrder) => void
}

export const SortingButtons: React.FC<SortingButtonsProps> = ({ onSort }) => {
  const [sortingOrder, setSortingOrder] = React.useState<SortingOrder>();

  const sortByAscending = () => {
    setSortingOrder(SortingOrder.ASC);
    onSort(SortingOrder.ASC);
  }

  const sortByDecending = () => {
    setSortingOrder(SortingOrder.DESC);
    onSort(SortingOrder.DESC);
  }

  return (
    <>
      <Button active={sortingOrder === SortingOrder.ASC} onClick={sortByAscending}>Sorter etter NÃ¦rmeste</Button>
      <Button active={sortingOrder === SortingOrder.DESC} onClick={sortByDecending}>Sorter etter Fjerneste</Button>
    </>
  );
}