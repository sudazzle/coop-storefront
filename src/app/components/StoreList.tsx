import React from 'react';
import { Address } from '@/app/icons/Address';
import { Phone } from '@/app/icons/Phone';
import { Store } from '@/app/types';

export const StoreOpenStatus: React.FC<{ open: boolean }> = ({ open }) => {
  return (
    <span className={
      `text-sm rounded-md py-1 px-3 ${open ? 'bg-green-400 text-white' : 'bg-red-500 text-red-600'}`
    }>
      {open ? 'Ã…pent' : 'Stengt'}
    </span>
  );
}

export const StoreCard: React.FC<{ store: Store }> = ({ store }) => {
  return (
    <li className="w-[90%] sm:w-[45%] md:w-[32%] rounded-2xl block p-5 bg-[#EFEFE9]">
      <h3 className="font-medium mb-3">{store.Chain}, {store.City}</h3>
      <p className="text-sm gap-1 text-gray-500 flex items-center">
        <span><Address /></span>
        <span>{ store.Address }</span>
      </p>
      <p className="text-sm gap-1 text-gray-500 flex items-center">
        <span><Phone /></span>
        <span>{ store.Phone }</span>
      </p>
      <p className="mt-5 text-sm">
        <StoreOpenStatus open={store.OpenNow} />
        <span className="pl-4 text-gray-500">{store.OpeningHoursToday}</span>
      </p>
    </li>
  );
}

export const StoreList: React.FC<{ stores: Store[] }> = ({ stores }) => {
  return (
    <ul className="flex mt-5 justify-center flex-wrap gap-4">
      {stores.map((store) => (
        <StoreCard key={store.StoreId} store={store} />
      ))}
    </ul>
  );
}