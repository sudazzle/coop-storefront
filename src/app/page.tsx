'use client';

import axios from 'axios';
import { Coordinate, haversineDistance } from '@/app/lib/distance';
import { SortingButtons } from './components/SortingButtons';
import { useEffect, useState } from 'react';
import { SortingOrder, Store, StoreByLocation } from '@/app/types';
import { StoreList } from './components/StoreList';

export const sortStoresByDistance = (stores: Store[], currentLocation: Coordinate, sortOrder: SortingOrder) => {
  const clone = [...stores];
  
  clone.sort((a, b) => {
    const distanceA = haversineDistance(currentLocation, {latitude: a.Lat, longitude: a.Lng});
    const distanceB = haversineDistance(currentLocation, {latitude: b.Lat, longitude: b.Lng});

    if (sortOrder === SortingOrder.ASC) {
      return distanceA - distanceB;
    } else {
      return distanceB - distanceA;
    }
  });

  return clone;
}

async function getData() {
  const res = await axios.get('/api/get-stores');
  return res.data as StoreByLocation;
}

export default function Home() {
  const [userLocation, setUserLocation] = useState<Coordinate>();
  const [stores, setStores] = useState<Store[]>([]);
  const [isLoading, setIsLoading] = useState(true);
  
  useEffect(() => {
    getData().then((data) => {
      setStores(data.Stores);

      if (data.Location) {
        setUserLocation({ latitude: data.Location.Latitude, longitude: data.Location.Longitude });
      }

    }).finally(() => {
      setIsLoading(false);
    });
  }, []);

  if (isLoading) {
    return <div className="text-center text-black p-5">Lasting...</div>;
  }

  const onSort = (sortOrder: SortingOrder) => {
    const sortedStores = sortStoresByDistance(stores, userLocation!, sortOrder);
    setStores(sortedStores);
  }

  return (
    <main>
      <h1 className="bg-slate-50 p-4 font-semibold text-center text-xl">VÃ¥re Butikker</h1>
      <div className="flex justify-center gap-3 p-4">
        {
          userLocation &&
          <SortingButtons onSort={onSort} />
        }
      </div>
      <StoreList stores={stores} />
    </main>
  );
}
