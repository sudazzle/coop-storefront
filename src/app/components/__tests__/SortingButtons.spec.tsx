import '@testing-library/jest-dom';
import { render, screen, fireEvent } from '@testing-library/react';
import { SortingButtons } from '@/app/components/SortingButtons';
import { SortingOrder } from '@/app/types';

describe('Page', () => {
  it('renders both sorting buttons', () => {
    const onSort = jest.fn();
    const { container } = render(<SortingButtons onSort={onSort} />)
    expect(screen.getAllByRole('button').length).toBe(2);
    expect(screen.getByText('Sorter etter Nærmeste')).toBeInTheDocument();
    expect(screen.getByText('Sorter etter Fjerneste')).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it('Sort by nearest should trigger sort event', () => {
    const onSort = jest.fn();
    render(<SortingButtons onSort={onSort} />)
    fireEvent.click(screen.getByText('Sorter etter Nærmeste'));
    expect(onSort).toHaveBeenCalledWith(SortingOrder.ASC);
  });

  it('Sort by farthest should trigger sort event', () => {
    const onSort = jest.fn();
    render(<SortingButtons onSort={onSort} />)
    fireEvent.click(screen.getByText('Sorter etter Fjerneste'));
    expect(onSort).toHaveBeenCalledWith(SortingOrder.DESC);
  });
});