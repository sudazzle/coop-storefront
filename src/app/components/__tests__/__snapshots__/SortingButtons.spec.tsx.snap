// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Page renders both sorting buttons 1`] = `
<div>
  <button
    class="hover:bg-blue-900 ease-in duration-300 hover:scale-105 hover:border-blue-900 hover:text-white border border-blue-300 rounded-xl py-2 px-4 bg-white text-blue-900"
  >
    Sorter etter NÃ¦rmeste
  </button>
  <button
    class="hover:bg-blue-900 ease-in duration-300 hover:scale-105 hover:border-blue-900 hover:text-white border border-blue-300 rounded-xl py-2 px-4 bg-white text-blue-900"
  >
    Sorter etter Fjerneste
  </button>
</div>
`;
